packages/artifact/src/internal-download-http-client.ts

```diff
@@ -12,122 +12,174 @@ import {
+      return JSON.parse(body)
+    }
+    // eslint-disable-next-line no-console
+    console.log(rawResponse)
+    throw new Error(`Unable to get ContainersItems from ${resourceUrl}`)
+  }
 
-/**
- * Downloads an individual file
- * @param client http client that will be used to make the necessary calls
- * @param artifactLocation origin location where a file will be downloaded from
- * @param downloadPath destination location for the file being downloaded
- */
-export async function downloadIndividualFile(
-  client: HttpClient,
-  artifactLocation: string,
-  downloadPath: string
-): Promise<void> {
-  const stream = fs.createWriteStream(downloadPath)
-  const response = await client.get(artifactLocation)
-  if (isSuccessStatusCode(response.message.statusCode)) {
-    await pipeResponseToStream(response, stream)
-  } else if (isRetryableStatusCode(response.message.statusCode)) {
-    warning(
-      `Received http ${response.message.statusCode} during file download, will retry ${artifactLocation} after 10 seconds`
+  /**
+   * Concurrently downloads all the files that are part of an artifact
+   * @param downloadItems information about what items to download and where to save them
+   */
+  async downloadSingleArtifact(downloadItems: DownloadItem[]): Promise<void> {
+    const DOWNLOAD_CONCURRENCY = getDownloadFileConcurrency()
+    // limit the number of files downloaded at a single time
+    const parallelDownloads = [...new Array(DOWNLOAD_CONCURRENCY).keys()]
+    this.downloadHttpManager.createClients(DOWNLOAD_CONCURRENCY)
+    let downloadedFiles = 0
+    await Promise.all(
+      parallelDownloads.map(async index => {
+        while (downloadedFiles < downloadItems.length) {
+          const currentFileToDownload = downloadItems[downloadedFiles]
+          downloadedFiles += 1
+          await this.downloadIndividualFile(
+            index,
+            currentFileToDownload.sourceLocation,
+            currentFileToDownload.targetPath
+          )
+        }
+      })
     )
-    await new Promise(resolve => setTimeout(resolve, 10000))
-    const retryResponse = await client.get(artifactLocation)
-    if (isSuccessStatusCode(retryResponse.message.statusCode)) {
-      await pipeResponseToStream(response, stream)
+
+    // done downloading, safety dispose all connections
+    this.downloadHttpManager.disposeAllConnections()
+  }
+
+  /**
+   * Downloads an individual file
+   * @param httpClientIndex the index of the http client that is used to make all of the calls
+   * @param artifactLocation origin location where a file will be downloaded from
+   * @param downloadPath destination location for the file being downloaded
+   */
+  private async downloadIndividualFile(
+    httpClientIndex: number,
+    artifactLocation: string,
+    downloadPath: string
+  ): Promise<void> {
+    const stream = fs.createWriteStream(downloadPath)
+    const client = this.downloadHttpManager.getClient(httpClientIndex)
```

Comment:
This client may be lost if the consumer called `listArtifacts()` or `getContainerItems()` again.

Reply:
See my above comment, `listArtifacts()`, much like `getContainerItems()` is an internal method and users won't be able to call it unless they do 
```
const downloadhttpclient = new DownloadHttpClient() // creates a new download http manager with its own group of clients
downloadhttpclient .listArtifacts()
```